crates/spirv-std/src/arch/ray_tracing.rs

```diff
@@ -0,0 +1,174 @@
+    // moment, we need to create and return the acceleration structure entirely
+    // in assembly.
+    asm! {
+        "%ret = OpTypeAccelerationStructureKHR",
+        "%result = OpConvertUToAccelerationStructureKHR %ret {id}",
+        "OpReturnValue %result",
+        "%blah = OpLabel",
+        id = in(reg) id,
+    }
+    loop {}
+}
+
+/// Trace a ray into the acceleration structure.
+///
+/// - `structure` is the descriptor for the acceleration structure to trace into.
+/// - `ray_flags` contains one or more of the Ray Flag values.
+/// - `cull_mask` is the mask to test against the instance mask. Only the 8
+///   least-significant bits of are used by this instruction - other bits
+///   are ignored.
+/// - `sbt_offset` and `sbt_stride` control indexing into the SBT (Shader
+///   Binding Table) for hit shaders called from this trace. Only the 4
+///   least-significant bits of `sbt_offset` and `sbt_stride` are used by this
+///   instruction - other bits are ignored.
+/// - `miss_index` is the index of the miss shader to be called from this
+///   trace call. Only the 16 least-significant bits are used by this
+///   instruction - other bits are ignored.
+/// - `ray_origin`, `ray_tmin`, `ray_direction`, and `ray_tmax` control the
+///   basic parameters of the ray to be traced.
+///
+/// - `payload` is a pointer to the ray payload structure to use for this trace.
+///   `payload` must have a storage class of `ray_payload_khr`
+///   or `incoming_ray_payload_khr`.
+///
+/// This instruction is allowed only in `ray_generation_khr`, `closest_hit_khr` and
+/// `miss_khr` execution models.
+///
+/// This instruction is a shader call instruction which may invoke shaders with
+/// the `intersection_khr`, `any_hit_khr`, `closest_hit_khr`, and `miss_khr`
+/// execution models.
+#[spirv_std_macros::gpu_only]
+#[doc(alias = "OpTraceRayKHR")]
+#[inline]
+#[allow(clippy::too_many_arguments)]
+pub unsafe fn trace_ray_khr<T>(
+    acceleration_structure: AccelerationStructureKhr,
+    ray_flags: i32,
+    cull_mask: i32,
+    sbt_offset: i32,
+    sbt_stride: i32,
+    miss_index: i32,
+    ray_origin: impl Vector<f32, 3>,
+    ray_tmin: f32,
+    ray_direction: impl Vector<f32, 3>,
+    ray_tmax: f32,
+    payload: &mut T,
+) {
+    asm! {
+        "OpTraceRayKHR \
+            {acceleration_structure} \
+            {ray_flags} \
+            {cull_mask} \
+            {sbt_offset} \
+            {sbt_stride} \
+            {miss_index} \
+            {ray_origin} \
+            {ray_tmin} \
+            {ray_direction} \
+            {ray_tmax} \
+            {payload}",
+        acceleration_structure = in(reg) &acceleration_structure,
+        ray_flags = in(reg) ray_flags,
+        cull_mask = in(reg) cull_mask,
+        sbt_offset = in(reg) sbt_offset,
+        sbt_stride = in(reg) sbt_stride,
+        miss_index = in(reg) miss_index,
+        ray_origin = in(reg) &ray_origin,
```

Comment:
... hold on, looking up the docs for this, `ray_origin` should be a straight vector, not a pointer to a vector, yet validation succeeds. I'm scared now, spirv-val probably doesn't validate this instruction at all, and we should probably throw this at an actual implementation to be sure?

Reply:
Oh yeah definitely, I don't think there's much validation of ray tracing instructions happening at all.